File: frontend/coordinator-android/app/src/main/java/ug/qaat/coordinator/ui/OfficeSearchMatch.kt
package ug.qaat.coordinator.ui

import ug.qaat.coordinator.db.OfficePersonEntity

/**
 * The OFFLINE office search, mirroring OfficeSearch on the server exactly.
 *
 * Extracted from the round screen for one reason: these two implementations diverging is a silent
 * failure. It does not throw and it does not log — it shows up only as a monitor standing at the
 * right door being told nobody by that name works here, and then either not filing the visit or
 * filing it against somebody else. So the rule lives in one function with a test on it rather than
 * inline in a composable nothing can exercise.
 *
 * THE RULE, and the SQL it mirrors:
 *
 *     AND ( btrim(lower(staff_id))                LIKE 'q%'     -- PREFIX
 *        OR btrim(lower(full_name))               LIKE '%q%'    -- CONTAINS
 *        OR btrim(lower(COALESCE(department,''))) LIKE '%q%'
 *        OR btrim(lower(COALESCE(office,'')))     LIKE '%q%' )
 *
 * PREFIX on the badge, because a monitor reads it off a card and types the start of it: "kiu/0"
 * must reach KIU/044. Deliberately NOT contains — "044" matching KIU/044 sounds helpful until a
 * three-digit fragment starts matching a third of the institution.
 *
 * CONTAINS on the name, department and office, because a registry name is "DR ANUMOLU SRINIVASA
 * RAO" and the nameplate on the door says ANUMOLU. Prefix-matching the name would return nothing
 * for the commonest thing a monitor types.
 */
internal object OfficeSearchMatch {

    fun matches(p: OfficePersonEntity, query: String): Boolean {
        val needle = query.trim().lowercase()
        // An empty query matches nobody, deliberately — the same rule the server applies, and the
        // same reason: a round that opens on everybody invites recording without visiting.
        if (needle.isEmpty()) return false

        return norm(p.staffId).startsWith(needle) ||
            norm(p.fullName).contains(needle) ||
            norm(p.department).contains(needle) ||
            norm(p.office).contains(needle)
    }

    fun filter(people: List<OfficePersonEntity>, query: String): List<OfficePersonEntity> =
        people.filter { matches(it, query) }.sortedBy { it.fullName.lowercase() }

    /** btrim(lower(COALESCE(x,''))) — the server normalises both sides, so this must too. */
    private fun norm(s: String?): String = (s ?: "").trim().lowercase()
}

File: frontend/coordinator-flutter/lib/patrol/office_search_match.dart
import '../patrol/patrol_models.dart';

/// The OFFLINE office search, mirroring the server's OfficeSearch exactly.
///
/// The rule lives in one function with a test on it rather than inline in a screen
/// nothing can exercise — the two implementations diverging is a silent failure: a
/// monitor standing at the right door is told nobody by that name works there.
///
/// Prefix-match the badge (a monitor types the start of "kiu/0" off a card, and
/// "044" as a contains-match would hit a third of the institution); contains-match
/// the name, department and office (a registry name is "DR ANUMOLU SRINIVASA RAO",
/// the nameplate says "ANUMOLU").
class OfficeSearchMatch {
  OfficeSearchMatch._();

  static bool matches(OfficePerson p, String query) {
    final needle = query.trim().toLowerCase();
    // An empty query matches nobody, deliberately — the server's rule; a round
    // that opens on everybody invites recording without visiting.
    if (needle.isEmpty) return false;
    return _norm(p.staffId).startsWith(needle) ||
        _norm(p.fullName).contains(needle) ||
        _norm(p.department).contains(needle) ||
        _norm(p.office).contains(needle);
  }

  static List<OfficePerson> filter(List<OfficePerson> people, String query) =>
      people.where((p) => matches(p, query)).toList()
        ..sort((a, b) => a.fullName.toLowerCase().compareTo(b.fullName.toLowerCase()));

  /// btrim(lower(COALESCE(x,''))) — the server normalises both sides, so this must too.
  static String _norm(String s) => s.trim().toLowerCase();
}